### PR TITLE
Rename "master" methods to "main" in ClusterEnvironment plugins

### LINKED ISSUE #8774: rename `master*` properties in cluster environments to `main*`
## 🚀 Feature

Refactor our internal language for [master port and master address]( https://github.com/PyTorchLightning/pytorch-lightning/blob/cc036155d482c9dbdc7a0d8a0be34696f314e192/pytorch_lightning/plugins/environments/cluster_environment.py#L29) in the cluster environments and accelerators.

### Motivation

Inclusive language

### Pitch

rename to `main_address` and `main_port`

### Alternatives

other name suggestions welcome

### Additional context

Lightning may also rename "master" branch to "main" soon.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>
``

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #8774
Continues #9094 (repository and branch was deleted)

In addition, converts the `main_port` and `main_address` methods to properties. This was requested in #6303. 

### Does your PR introduce any breaking changes? If yes, please list them.

Yes. User with custom cluster environment plugins won't be able to instantiate their existing plugins due to the requirement of `main_port`, `main_address` to be abstract. They will have to rename `master_port` and `master_address` methods immediately. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

### CODE REVIEW

File: pytorch_lightning/plugins/environments/cluster_environment.py [hunk 1]
```diff
@@ -72,3 +79,16 @@ def node_rank(self) -> int:
     def teardown(self) -> None:
         """Clean up any state set after execution finishes."""
         pass
+
+
+def _check_for_deprecated_methods(cls: Type[ClusterEnvironment]) -> None:
+    if hasattr(cls, "master_address") and callable(cls.master_address):
+        rank_zero_deprecation(
+            f"`{cls.__name__}.master_address` has been deprecated in v1.6 and will be removed in 1.7."
+            " Implement the property `main_address` instead (do not forget to add the `@property` decorator."
```
Comment (justusschock) [inline pytorch_lightning/plugins/environments/cluster_environment.py:88]:
```suggestion
            " Implement the property `main_address` instead (do not forget to add the `@property` decorator)."
```

File: pytorch_lightning/plugins/environments/cluster_environment.py [hunk 2]
```diff
@@ -72,3 +79,16 @@ def node_rank(self) -> int:
...
+        )
+    if hasattr(cls, "master_port") and callable(cls.master_port):
+        rank_zero_deprecation(
+            f"`{cls.__name__}.master_port` has been deprecated in v1.6 and will be removed in 1.7."
+            " Implement the property `main_port` instead (do not forget to add the `@property` decorator."
```
Comment (justusschock) [inline pytorch_lightning/plugins/environments/cluster_environment.py:93]:
```suggestion
            " Implement the property `main_port` instead (do not forget to add the `@property` decorator)."
```

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/plugins/environments/cluster_environment.py [hunk 3]
```diff
@@ -37,13 +43,15 @@ def creates_children(self) -> bool:
         )
         return self.creates_processes_externally
 
+    @property
```
Comment (carmocca) [inline pytorch_lightning/plugins/environments/cluster_environment.py:46]: You are using this PR to add these missed property decorators. Is this something that should go into the bug-fix release? If yes, you'll need to split this PR

File: pytorch_lightning/plugins/environments/lsf_environment.py [hunk 4]
```diff
@@ -56,13 +56,15 @@ def is_using_lsf() -> bool:
     def creates_processes_externally(self) -> bool:
         return True
 
-    def master_address(self):
-        """The master address is read from a list of hosts contained in the environment variable `LSB_HOSTS`."""
-        return self._master_address
+    @property
+    def main_address(self):
```
Comment (carmocca) [inline pytorch_lightning/plugins/environments/lsf_environment.py:60]: nit: add return typing, same for other modified occurrences in this file

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/plugins/environments/cluster_environment.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/environments/cluster_environment.py:46]: No, I don't think I'm adding any missing ones.
They never were properties. I'm combining the renaming + the property change, as it is a breaking change anyway.

The PR title is misleading, changing it.

File: pytorch_lightning/plugins/environments/lsf_environment.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/environments/lsf_environment.py:60]: ah good call, just realized we don't have mypy for cluster envs. should be easy to do.

File: pytorch_lightning/plugins/environments/cluster_environment.py (hunk 3 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/environments/cluster_environment.py:46]: should these be properties or static methods?

File: pytorch_lightning/plugins/environments/lsf_environment.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/environments/lsf_environment.py]: added types but need to send PR to properly make mypy pass on some methods

File: pytorch_lightning/plugins/environments/cluster_environment.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/environments/cluster_environment.py:46]: there are plugins like the LSF that save some state and read it. So static would not work there.
I think if we want these plugins be customizable, non-static is the most flexible. I don't think there is a cost associated with it not to do it. 

Property or method, I don't know, don't care. I only do it because when I added a set_x() method in this PR #6941, people requested me to convert it to properties and setters (https://github.com/PyTorchLightning/pytorch-lightning/pull/6941#pullrequestreview-633644884). Then I created this issue here #6303.

File: docs/source/clouds/cluster.rst [hunk 5]
```diff
@@ -82,7 +82,7 @@ Once the script is setup like described in :ref:`training_script_setup`, you can
 
 Like a custom cluster, you have to ensure that there is network connectivity between the nodes with firewall rules that allow traffic flow on a specified *MASTER_PORT*.
 
-Finally, you'll need to decide which node you'd like to be the master node (*MASTER_ADDR*), and the ranks of each node (*NODE_RANK*).
+Finally, you'll need to decide which node you'd like to be the main node (*MASTER_ADDR*), and the ranks of each node (*NODE_RANK*).
```
Comment (Borda) [inline docs/source/clouds/cluster.rst:85]:
```suggestion
Finally, you'll need to decide which node you'd like to be the main node (*MAIN_ADDR*), and the ranks of each node (*NODE_RANK*).
```
shall this be also main?

File: pytorch_lightning/plugins/environments/cluster_environment.py [hunk 6]
```diff
@@ -12,23 +12,30 @@
...
 from abc import ABC, abstractmethod
+from typing import Any, Type
 
 
 class ClusterEnvironment(ABC):
     """Specification of a cluster environment."""
 
+    def __new__(cls, *args: Any, **kwargs: Any) -> "ClusterEnvironment":
+        # TODO: remove in 1.7
+        _check_for_deprecated_methods(cls)
```
Comment (Borda) [inline pytorch_lightning/plugins/environments/cluster_environment.py:23]: just check? not filling the old master_ to new main_

File: pytorch_lightning/plugins/environments/lightning_environment.py [hunk 7]
```diff
@@ -49,13 +49,15 @@ def creates_processes_externally(self) -> bool:
         """
         return "LOCAL_RANK" in os.environ
 
-    def master_address(self) -> str:
+    @property
+    def main_address(self) -> str:
```
Comment (Borda) [inline pytorch_lightning/plugins/environments/lightning_environment.py:53]: feel this is just adding confusion when we rename MASTER_ from os env to main_ locally

**Review by Borda — COMMENTED** [review]
feeling it is just adding confusion if we preserve MASTER_ in OS environment and locally in PL rename as main_ can we do the rename for both?

**Comment by awaelchli** [comment]
To clarify, @Borda the MASTER_* variables in os.environ are set by pytorch and read by pytorch (or rather the distributed backends). We can't rename them, but it's also not a problem because these are not user facing. However, our public API for the cluster plugins is and I think we should update them. Another example was #10105.

File: pytorch_lightning/plugins/environments/cluster_environment.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/environments/cluster_environment.py:25]: I think it can't be done so easily, because one is a property and the other is a method. I'm not sure what the best is actually. Since the new properties are also abstract, the user is forced to implement them. So a deprecation message is maybe not appropriate. Should it be a direct error?

Note: This only concerns people implementing a custom cluster plugin. Anyone using an existing one is fine.

Comment (tchaton) [reply pytorch_lightning/plugins/environments/cluster_environment.py:25]: Neat ! I like this design. Might be worth to automate this for most classes :)

**Review by tchaton — APPROVED** [review]
LGTM !
